File: algorithms/excelSheetColumnNumber/excelSheetColumnNumber.js
/****
 * Solution of https://leetcode.com/problems/excel-sheet-column-number/
 * @author: Jerry Zou
 * @date: 2015.6.4
 **/

/**
 * @param {string} s
 * @return {number}
 */
var titleToNumber = function(s) {
    var column = 0,
        i = s.length-1,
        base = 1,
        map = {
            A: 1, B: 2, C: 3, D: 4,
            E: 5, F: 6, G: 7, H: 8,
            I: 9, J: 10, K: 11, L: 12,
            M: 13, N: 14, O: 15, P: 16,
            Q: 17, R: 18, S: 19, T: 20,
            U: 21, V: 22, W: 23, X: 24,
            Y: 25, Z: 26
        };

    for (; i >= 0; i--) {
        column += map[s[i]] * base;
        base *= 26;
    }

    return column;
};

File: algorithms/java/src/ExcelSheetColumnNumber/ExcelSheetColumnNumber.java
// Source : https://leetcode.com/problems/excel-sheet-column-number/
// Author : Diego Ruiz Piqueras (Pikeras72)
// Date   : 24-04-2022

/***************************************************************************************************** 
 * Given a string columnTitle that represents the column title as appear in an
 * Excel sheet, return its corresponding column number.
 * 
 * For example:
 * A -> 1
 * B -> 2
 * C -> 3
 * ...
 * Z -> 26
 * AA -> 27
 * AB -> 28 
 * ...
 * 
 * Example 1:
 * 
 * Input: columnTitle = "A"
 * Output: 1
 * 
 * Example 2:
 *
 * Input: columnTitle = "AB"
 * Output: 28
 * Explanation: 
 * 
 * Example 3:
 * 
 * Input: columnTitle = "ZY"
 * Output: 701
 * 
 *  1 <= columnTitle.length <= 7
 *  columnTitle consists only of uppercase English letters.
 * columnTitle is in the range ["A", "FXSHRXW"].
 * 
 * Explanation of the solution:
 * 
 * 1. While we have given less candies than the ones we had at the beginning (while(total < candies){...})
 * 
 * 2. Use count % num_people == 0 to determine the current index of the people.
 ******************************************************************************************************/

class Solution {
    public int titleToNumber(String columnTitle) {
        int res = 0, cnt = 0;
        String letras = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
        int exponente = columnTitle.length()-1;
        if(exponente == 0){
            return letras.indexOf(columnTitle)+1;
        }
        while(exponente != -1){
            res += (int) ((letras.indexOf(String.valueOf(columnTitle.charAt(cnt)))+1)*Math.pow(26, exponente));
            exponente--;
            cnt++;
        }
        return res;
    }
}
